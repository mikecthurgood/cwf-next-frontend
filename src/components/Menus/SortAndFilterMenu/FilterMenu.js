import React, {useState} from 'react';
import Styled from 'styled-components';

import WallFilter  from './WallFilter';
import PostcodeSort  from './PostcodeSort';

const FilterMenu = () => {
    const [filterMenuVisible, setFilterMenuVisibility] = useState(false)

    function toggleMenuVisibility () {
        setFilterMenuVisibility(!filterMenuVisible)
    }

    return (
        <>
        <BackgroundFade className={filterMenuVisible ? 'visible' : ''} onClick={toggleMenuVisibility}></BackgroundFade>
        <FilterMenuContainer>
            <label className='filter__menu-title' onClick={toggleMenuVisibility}>
                <label className='filter__menu-icon'>
                        <strong>{`${filterMenuVisible ? '-' : '+'}`}</strong>
                </label>
                    <strong>Sort & Filter</strong>
            </label>
            <div data-testid='filter-menu' className={`filter__menu-filter-container ${filterMenuVisible ? 'visible' : ''}`} >
                <div className='filter__menu-filter-item' >
                    <WallFilter />
                </div>
                <div className='filter__menu-filter-item' >
                    <PostcodeSort setFilterMenuVisibility={setFilterMenuVisibility} />
                </div>
            </div>
        </FilterMenuContainer>
        </>
    )
}

export default FilterMenu

const BackgroundFade = Styled.div`
    opacity: 0;
    width: 100vw;
    height: 100vh;
    background-color: black;
    z-index: -1000;
    position: fixed;
    top: 40px;
    transition: opacity 0.45s ease-in-out;
    visibility: hidden;
    &.visible {
        display: block;
        z-index: 10;
        opacity: 0.8;
        visibility: visible;
        transition: opacity 0.45s ease-in-out;
    }
`

const FilterMenuContainer = Styled.div`
    z-index: 101;
    position: fixed;
    text-align: left;
    display: flex;
    left: 0;
    top: 3.7rem;
    color: white;
    padding-left: 1rem;
    width: 25%;

    .filter__menu-mobile-heading, .filter__menu-icon {
        display: none;
    }
    .filter__menu-title {
        display: flex;
        flex-direction: row;
        text-transform: uppercase;

    }
    
    .filter__menu-icon {
        position: relative;
        left: 0;
        margin-right: 5px;
        font-size: 30px;
        bottom: 11px;
        display: flex;
    }
    .filter__menu-form-heading {
        display: flex;
        padding-left: 20px;
    }
    .filter__menu-filter-container {
        position: absolute;
        transform: translateX(-100vw);
        height: 100vh;
        top: 30px;
        left: -22px;
        padding: 20px;
        background-color: #ffffff;
        color: #282c34;
        min-width: 250px;
        opacity: 0.9;
        transition: transform 0.45s ease-in-out;
        &.visible {
            transform: translateX(0);
            transition: transform 0.45s ease-in-out;
        }

        .filter__menu-filter-item {
            margin-bottom: 15px;
        }
    }

    @media only screen and (max-width: 480px) {
        position: absolute;
        left: 0;
        width: 50%;
        padding-left: 5px;
        
        .filter__menu-filter-container {
            width: 50vw;
            left: 0;
            min-width: 200px;
            .filter__menu-filter-item {

                .filter__menu-form-heading {
                    padding-left: 5px;
                }
                form {
                    .filter__menu-checkbox {
                        padding-left: 5px;
                        width: 100%;
                        margin-bottom: 10px;
                    }
                }
                .filter__menu-sort-form {
                    .filter__menu-sort-input {
                        left: 0;
                    }
                    .submit-postcode {
                        right: -158px;
                    }
                }
                .change-post-code-container {
                    padding-left: 5px;
                }
            }
        }
    }
`