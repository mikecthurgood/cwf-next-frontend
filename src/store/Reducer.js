const reducer = (state, action) => {
    switch (action.type) {
        
        case 'setFilterSelection':
            return { ...state, filterSelection: action.data}
        case 'setLoginError':
            return { ...state, loginError: action.data}
        case 'setLoginMenuVisible':
            return { ...state, loginMenuVisible: action.data}
        case 'setMobileMenuVisible':
            return { ...state, mobileMenuVisible: action.data}
        case 'setSearchBarVisibility':
            return { ...state, searchBarVisible: action.data}
        case 'setSearchFilter':
            return { ...state, searchFilter: action.data}
        case 'setScrollPosition':
            return { ...state, scrollPosition: action.data}
        case 'setSignUpSuccess':
            return { ...state, signUpSuccess: action.data}
        case 'setSingleWall':
            return { ...state, singleWall: action.data}
        case 'setSortInputVisibility':
            return { ...state, sortInputVisible: action.data}
        case 'setUser':
            return { ...state, user: action.data}
        case 'setPostCode':
            return { ...state, userPostCode: action.data}
        case 'setWalls':
            return { ...state, walls: action.data}
        case 'signOut' :
            localStorage.removeItem('token')
            localStorage.removeItem('userName')
            localStorage.removeItem('userId')
            return { 
                ...state,
                user: {username: null, userId: null, isAuth: false, token: ''},
                mobileMenuVisible: false,
                loginMenuVisible: false
            }
        default: 
            throw new Error('Invalid action type dispatched')
    }
}
export default reducer