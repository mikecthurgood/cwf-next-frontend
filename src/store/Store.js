import React, { useReducer } from  'react';
import reducer from './Reducer'

const initialState = {
    filterSelection: [],
    loginError: false,
    loginMenuVisible: false,
    mobileMenuVisible: false,
    searchBarVisible: false,
    scrollPosition: 0,
    signUpSuccess: false,
    singleWall: {},
    sortInputVisible: false,
    user: { username: null, userId: null, isAuth: false, token: '' },
    userPostCode: '',
    walls: [],
}

const State = React.createContext(initialState)
const Dispatch = React.createContext(null)

const Store = ({children}) => {
    
    const [state, dispatch] = useReducer(reducer, initialState)

    return (
        <Dispatch.Provider value={dispatch} >
            <State.Provider value={state} >
                {children}
            </State.Provider>
        </Dispatch.Provider>
    )
}


export { State, Dispatch, Store }

