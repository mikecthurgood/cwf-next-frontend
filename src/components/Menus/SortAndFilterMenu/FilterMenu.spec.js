import React from 'react';
import { cleanup, fireEvent, render } from '@testing-library/react';
import { Dispatch, State } from '../../../store/Store';
import '@testing-library/jest-dom';
import FilterMenu from '.';


describe('Filter Menu', () => {
    let dispatch;
    beforeEach(() => {
        dispatch = jest.fn();
    })

    afterEach(() => {
        cleanup();
    })
    const state = {
        filterSelection: []
    };

    it('should render the component', () => {
        const { container } = render(
            <Dispatch.Provider value={dispatch}>
                <State.Provider value={state}>
                    <FilterMenu />
                </State.Provider>
            </Dispatch.Provider>
        );
        expect(container).toMatchSnapshot();
    });

    it('should set the menu to visible on click', () => {
        const { container, getByText, getByTestId } = render(
            <Dispatch.Provider value={dispatch}>
                <State.Provider value={state}>
                    <FilterMenu />
                </State.Provider>
            </Dispatch.Provider>
        );
        const menuToggle = getByText('+');
        fireEvent.click(menuToggle);
        expect(getByTestId('filter-menu').classList).toContain('visible')
        expect(container).toMatchSnapshot();
    })

    it('should allow users to select a filter via a checkbox', () => {
        const { getByText } = render(
            <Dispatch.Provider value={dispatch}>
                <State.Provider value={state}>
                    <FilterMenu />
                </State.Provider>
            </Dispatch.Provider>
        );
        const autoBelayFilter = getByText('Auto Belay:')

        fireEvent.click(autoBelayFilter);

        expect(dispatch).toHaveBeenCalledWith({
            type: 'setFilterSelection',
            data: ['auto'],
        })
    })

    it('should allow users to select an additional filter when one is already selected', () => {
        state.filterSelection = ['auto']
        const { getByText } = render(
            <Dispatch.Provider value={dispatch}>
                <State.Provider value={state}>
                    <FilterMenu />
                </State.Provider>
            </Dispatch.Provider>
        );
        const boulderingFilter = getByText('Bouldering:')
        fireEvent.click(boulderingFilter);

        expect(dispatch).toHaveBeenCalledWith({
            type: 'setFilterSelection',
            data: ['auto', 'bouldering'],
        })
    })

    it('should allow users to deselect a filter if it is already selected', () => {
        state.filterSelection = ['auto', 'top']
        const { getByText } = render(
            <Dispatch.Provider value={dispatch}>
                <State.Provider value={state}>
                    <FilterMenu />
                </State.Provider>
            </Dispatch.Provider>
        );
        const topRopeFilter = getByText('Top Roping:')
        fireEvent.click(topRopeFilter);

        expect(dispatch).toHaveBeenCalledWith({
            type: 'setFilterSelection',
            data: ['auto'],
        })
    })
})