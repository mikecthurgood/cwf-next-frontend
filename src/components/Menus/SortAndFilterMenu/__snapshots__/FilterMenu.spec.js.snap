// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Menu should render the component 1`] = `
<div>
  <div
    class="FilterMenu__BackgroundFade-sc-1fra4qv-0 jTtCeR"
  />
  <div
    class="FilterMenu__FilterMenuContainer-sc-1fra4qv-1 cMfzVK"
  >
    <label
      class="filter__menu-title"
    >
      <label
        class="filter__menu-icon"
      >
        <strong>
          +
        </strong>
      </label>
      <strong>
        Sort & Filter
      </strong>
    </label>
    <div
      class="filter__menu-filter-container "
      data-testid="filter-menu"
    >
      <div
        class="filter__menu-filter-item"
      >
        <div
          class="WallFilter__WallFilterContainer-sc-1gmh8t7-0 dlCtk"
        >
          <label
            class="filter__menu-form-heading"
          >
            <strong>
              Wall Type
            </strong>
          </label>
          <form>
            <label
              class="filter__menu-checkbox"
            >
              Bouldering:
              <input
                name="bouldering"
                type="checkbox"
              />
              <span
                class="checkbox"
              />
            </label>
            <label
              class="filter__menu-checkbox"
            >
              Top Roping:
              <input
                name="top"
                type="checkbox"
              />
              <span
                class="checkbox"
              />
            </label>
            <label
              class="filter__menu-checkbox"
            >
              Lead:
              <input
                name="lead"
                type="checkbox"
              />
              <span
                class="checkbox"
              />
            </label>
            <label
              class="filter__menu-checkbox"
            >
              Auto Belay:
              <input
                name="auto"
                type="checkbox"
              />
              <span
                class="checkbox"
              />
            </label>
          </form>
        </div>
      </div>
      <div
        class="filter__menu-filter-item"
      >
        <div
          class="PostcodeSort__PostcodeSortContainer-kmt73j-0 fmYhrI"
        >
          <label
            class="filter__menu-form-heading"
          >
            <strong>
              Sort by distance
            </strong>
          </label>
          <div
            class="change-post-code-container"
          >
            <form
              autocomplete="off"
              class="filter__menu-sort-form visible"
            >
              <input
                class="filter__menu-sort-input"
                name="searchInput"
                placeholder="Enter Postcode"
                type="text"
                value=""
              />
              <input
                class="PostcodeSort__SubmitButton-kmt73j-1 bfOvYG"
                type="submit"
                value="Set"
              />
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Filter Menu should set the menu to visible on click 1`] = `
<div>
  <div
    class="FilterMenu__BackgroundFade-sc-1fra4qv-0 jTtCeR visible"
  />
  <div
    class="FilterMenu__FilterMenuContainer-sc-1fra4qv-1 cMfzVK"
  >
    <label
      class="filter__menu-title"
    >
      <label
        class="filter__menu-icon"
      >
        <strong>
          -
        </strong>
      </label>
      <strong>
        Sort & Filter
      </strong>
    </label>
    <div
      class="filter__menu-filter-container visible"
      data-testid="filter-menu"
    >
      <div
        class="filter__menu-filter-item"
      >
        <div
          class="WallFilter__WallFilterContainer-sc-1gmh8t7-0 dlCtk"
        >
          <label
            class="filter__menu-form-heading"
          >
            <strong>
              Wall Type
            </strong>
          </label>
          <form>
            <label
              class="filter__menu-checkbox"
            >
              Bouldering:
              <input
                name="bouldering"
                type="checkbox"
              />
              <span
                class="checkbox"
              />
            </label>
            <label
              class="filter__menu-checkbox"
            >
              Top Roping:
              <input
                name="top"
                type="checkbox"
              />
              <span
                class="checkbox"
              />
            </label>
            <label
              class="filter__menu-checkbox"
            >
              Lead:
              <input
                name="lead"
                type="checkbox"
              />
              <span
                class="checkbox"
              />
            </label>
            <label
              class="filter__menu-checkbox"
            >
              Auto Belay:
              <input
                name="auto"
                type="checkbox"
              />
              <span
                class="checkbox"
              />
            </label>
          </form>
        </div>
      </div>
      <div
        class="filter__menu-filter-item"
      >
        <div
          class="PostcodeSort__PostcodeSortContainer-kmt73j-0 fmYhrI"
        >
          <label
            class="filter__menu-form-heading"
          >
            <strong>
              Sort by distance
            </strong>
          </label>
          <div
            class="change-post-code-container"
          >
            <form
              autocomplete="off"
              class="filter__menu-sort-form visible"
            >
              <input
                class="filter__menu-sort-input"
                name="searchInput"
                placeholder="Enter Postcode"
                type="text"
                value=""
              />
              <input
                class="PostcodeSort__SubmitButton-kmt73j-1 bfOvYG"
                type="submit"
                value="Set"
              />
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
